docs/proposals/scheduling/20220530-fine-grained-cpu-orchestration.md

```diff
@@ -0,0 +1,620 @@
+            - [Reserve phase](#reserve-phase)
+            - [PreBind phase](#prebind-phase)
+            - [CPU Allocation Algorithm](#cpu-allocation-algorithm)
+            - [Plugin Configuration](#plugin-configuration)
+    - [Alternatives](#alternatives)
+        - [Defined Koordinator NodeResourceTopology CRD](#defined-koordinator-noderesourcetopology-crd)
+    - [Unsolved Problems](#unsolved-problems)
+    - [Implementation History](#implementation-history)
+
+<!-- /TOC -->
+
+# Fine-grained CPU orchestration
+
+## Summary
+
+This Proposal defines the fine-grained CPU orchestration for Koordinator QoS in detail, and how to be compatible with the existing design principles and implementations of K8s. This proposal describes the functionality that koordlet, koord-runtime-proxy and koord-scheduler need to enhance.
+
+## Motivation
+An increasing number of systems leverage a combination of CPUs and hardware accelerators to support latency-critical execution and high-throughput parallel computation. These include workloads in fields such as telecommunications, scientific computing, machine learning, financial services and data analytics. Such hybrid systems comprise a high performance environment.
+In order to extract the best performance,  optimizations related to CPU isolation,  NUMA-locality are required.  
+
+### Goals
+
+1. Improve the CPU orchestration semantics of Koordinator QoS.
+1. Determine compatible kubelet policies.
+1. Clarified that koordlet should enhance CPU scheduling mechanism.
+1. Provide a set of API such as CPU bind strategies, CPU isolate strategies, NUMA topology alignment strategies, NUMA resources information etc. for applications and cluster administrator to support complex CPU orchestration scenarios.
+1. Provide the CPU orchestration optimization API.
+
+### Non-Goals/Future Work
+
+1. Describe specific design details of koordlet/koord-runtime-proxy.
+1. Describe specific design details of CPU descheduling mechanism.
+
+## Proposal
+
+### Design Overview
+
+![image](/docs/images/cpu-orchestration-seq-uml.svg)
+
+When koordlet starts, koordlet gather the NUMA topology informations from kubelet include NUMA Topology, CPU Topology, kubelet cpu management policy, kubelet allocated CPUs for Guarantee Pods etc. , and update to the NodeResourceTopology CRD.  The latency-sensitive applications are scaling, the new Pod can set Koordinator QoS with LSE/LSR, CPU Bind strategy and CPU Isolate strategy to require koord-scheduler to allocate best-fit CPUs to get the best performance. When koord-scheduler schedling the Pod, koord-scheduler will filter Nodes that satisifed NUMA Topology Alignment strategy, and select the best Node by scoring, allocating the CPUs in Reserve phase, and records the result to Pod annotation when PreBinding. koordlet hooks the kubelet CRI request to replaces the CPUs configuration paramters with the koord-scheduler scheduled result to the runtime such as configurate the cgroup.
```

Comment:
When koord-scheduler schedling[scheduling] the Pod

Reply:
done